### PR TITLE
Support big page PUT into blobstore.

### PR DESCRIPTION
**Pull request by jiaqizho**
### What problem does this PR solve?

Issue Number: ref #3594 

Problem Summary:
- `BlobStore` can't support a single page whose page size is bigger than `config.file_limit_size`.
- Also can't support `WriteBatch` whose size is bigger than `config.file_limit_size`.

### What is changed and how it works?
- Added a new `SpaceMap` type which named `SMAP64_BIG`
  - It pairs with page which sizes bigger than `config.file_limit_size`
  - This type of `SpaceMap` can't be insert twice and also can't be reused.
- Added a new BlobStatType named `BIG_BLOB`
  - This type of stat only can use `SMAP64_BIG` type to store the address.
  - And this type of stat won't be GC but it can be deleted if that page has been removed.
  - this type of stat won't be reused.
- Also split big size of `WriteBatch` in `write()`.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/Page/V3/spacemap/SpaceMapBig.h [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+namespace DB
+{
+namespace ErrorCodes
+{
+extern const int LOGICAL_ERROR;
+}
+
+namespace PS::V3
+{
+class BigSpaceMap
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/spacemap/SpaceMapBig.h:33]:
```suggestion
// A space map that is designed for holding only one large page data (size > blobstore.config.file_limit_size)
class BigSpaceMap
```

File: dbms/src/Storages/Page/V3/BlobStore.cpp [hunk 2]
```diff
@@ -101,20 +109,100 @@ void BlobStore::registerPaths()
...
 
-    PageEntriesEdit edit;
     const size_t all_page_data_size = wb.getTotalDataSize();
 
     if (all_page_data_size > config.file_limit_size)
     {
-        throw Exception(fmt::format("Write batch is too large. It should less than [file_limit_size={}]",
-                                    config.file_limit_size.get()),
-                        ErrorCodes::LOGICAL_ERROR);
+        return splitWrite(wb, write_limiter);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/BlobStore.cpp:201]: I think name it "handleLargeWrite" is better

File: dbms/src/Storages/Page/V3/BlobStore.h [hunk 3]
```diff
@@ -126,6 +137,16 @@ class BlobStore : private Allocator<false>
...
 
+            bool isBigBlob() const
+            {
+                return type.load() == BlobStatType::BIG_BLOB;
+            }
+
+            void changeToBigBlob()
+            {
+                type.store(BlobStatType::BIG_BLOB);
+            }
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/BlobStore.h:148]: can we set it in the ctor of `BlobStat` when `sm_type==big`?

File: dbms/src/Storages/Page/V3/BlobStore.cpp [hunk 4]
```diff
@@ -90,8 +90,16 @@ void BlobStore::registerPaths()
             if (blob_id != INVALID_BLOBFILE_ID)
             {
                 Poco::File blob(fmt::format("{}/{}", path, blob_name));
-                delegator->addPageFileUsedSize({blob_id, 0}, blob.getSize(), path, true);
-                blob_stats.createStatNotChecking(blob_id, lock_stats);
+                auto blob_size = blob.getSize();
+                delegator->addPageFileUsedSize({blob_id, 0}, blob_size, path, true);
+                if (blob_size > config.file_limit_size)
+                {
+                    blob_stats.createBigStat(blob_id, lock_stats);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/BlobStore.cpp:97]: seems it will move `roll_id` forward and make BlobStore not the same as the last?

Comment (JaySon-Huang) [reply dbms/src/Storages/Page/V3/BlobStore.cpp:97]: make it behavior like `createBigStatNotChecking`

File: dbms/src/Storages/Page/V3/BlobStore.cpp [hunk 5]
```diff
@@ -1140,6 +1222,27 @@ BlobStatPtr BlobStore::BlobStats::createStatNotChecking(BlobFileId blob_file_id,
     return stat;
 }
 
+BlobStatPtr BlobStore::BlobStats::createBigStat(BlobFileId blob_file_id, const std::lock_guard<std::mutex> &)
+{
+    LOG_FMT_DEBUG(log, "Created a new big BlobStat [blob_id={}]", blob_file_id);
```
[the remark this answers is not in the document]
Comment (jiaqizho) [reply dbms/src/Storages/Page/V3/BlobStore.cpp:1227]: Changed. also changed Log in `createStatNotChecking`.

**Review by hehechen — COMMENTED** [review]
Add description about `What is changed and how it works?` and  update the design doc.

File: dbms/src/Storages/Page/V3/BlobStore.h (hunk 3 above)
Comment (jiaqizho) [reply dbms/src/Storages/Page/V3/BlobStore.h:148]: what is `ctor`? i don't get it.

File: dbms/src/Storages/Page/V3/BlobStore.cpp (hunk 4 above)
Comment (jiaqizho) [reply dbms/src/Storages/Page/V3/BlobStore.cpp:97]: split it into `createBigStat` and `createBigPageStatNotChecking`.

File: dbms/src/Storages/Page/V3/BlobStore.h (hunk 3 above)
Comment (JaySon-Huang) [reply dbms/src/Storages/Page/V3/BlobStore.h:148]: `BlobStat::BlobStat(...)`

**Comment by jiaqizho** [comment]
> Add description about `What is changed and how it works?` and update the design doc.

done

**Review by hehechen — APPROVED** [review]
Add description about `What is changed and how it works?` and  update the design doc.

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Pull request merged**